File: worker-python/src/modules/queue/job_ids.py
from __future__ import annotations

from collections.abc import Iterable


JOB_ID_MIN_WIDTH = 4


def format_job_id(sequence_number: int) -> str:
    if sequence_number <= 0:
        raise ValueError("sequence_number must be greater than 0")

    return str(sequence_number).zfill(JOB_ID_MIN_WIDTH)


def parse_job_id(job_id: str) -> int:
    normalized_job_id = job_id.strip()
    if normalized_job_id == "":
        raise ValueError("job_id is required")
    if not normalized_job_id.isdigit():
        raise ValueError("job_id must contain only digits")

    parsed = int(normalized_job_id)
    if parsed <= 0:
        raise ValueError("job_id must be greater than 0")

    return parsed


def get_next_job_id(existing_job_ids: Iterable[str]) -> str:
    highest_sequence = 0
    for existing_job_id in existing_job_ids:
        highest_sequence = max(highest_sequence, parse_job_id(existing_job_id))

    return format_job_id(highest_sequence + 1)

File: worker-node/src/modules/queue/jobIds.ts
const JOB_ID_MIN_WIDTH = 4;

export const formatJobId = (sequenceNumber: number): string => {
  if (sequenceNumber <= 0) {
    throw new Error('sequenceNumber must be greater than 0');
  }

  return String(sequenceNumber).padStart(JOB_ID_MIN_WIDTH, '0');
};

export const parseJobId = (jobId: string): number => {
  const normalized = jobId.trim();
  if (normalized === '') {
    throw new Error('jobId is required');
  }
  if (!/^\d+$/.test(normalized)) {
    throw new Error('jobId must contain only digits');
  }

  const parsed = parseInt(normalized, 10);
  if (parsed <= 0) {
    throw new Error('jobId must be greater than 0');
  }

  return parsed;
};

export const getNextJobId = (existingJobIds: string[]): string => {
  let highestSequence = 0;

  for (const existingJobId of existingJobIds) {
    highestSequence = Math.max(highestSequence, parseJobId(existingJobId));
  }

  return formatJobId(highestSequence + 1);
};
